### PR TITLE
 Added Multi Faction Support to Doors 

### PR DESCRIPTION
**Pull request by bleonheart**
Adds Several Commands Requested By People to Add Multiple Factions to Doors

### CODE REVIEW

File: plugins/doors/entities/weapons/nut_keys.lua [hunk 1]
```diff
@@ -1,164 +1,126 @@
...
-	SWEP.Slot = 0
-	SWEP.SlotPos = 2
-	SWEP.DrawAmmo = false
+if CLIENT then
+    SWEP.PrintName = "Keys"
+    SWEP.Slot = 0
+    SWEP.SlotPos = 2
+    SWEP.DrawAmmo = false
+    SWEP.DrawCrosshair = false
 end
```
Comment (zoephix) [inline plugins/doors/entities/weapons/nut_keys.lua:9]: Unnecessary changes.

File: plugins/doors/entities/weapons/nut_keys.lua [hunk 2]
```diff
@@ -1,164 +1,126 @@
...
 
 SWEP.Author = "Chessnut"
-SWEP.Instructions = "Primary Fire: Lock\nSecondary Fire: Unlock"
+SWEP.Instructions = "Primary Fire: Lock Entities & Drag Players\nSecondary Fire: Unlock"
```
Comment (zoephix) [inline plugins/doors/entities/weapons/nut_keys.lua:12]: again

File: plugins/doors/entities/weapons/nut_keys.lua [hunk 3]
```diff
@@ -1,164 +1,126 @@
...
 SWEP.Purpose = "Hitting things and knocking on doors."
 SWEP.Drop = false
-
 SWEP.ViewModelFOV = 45
 SWEP.ViewModelFlip = false
-SWEP.AnimPrefix	 = "passive"
-
+SWEP.AnimPrefix = "rpg"
```
Comment (zoephix) [inline plugins/doors/entities/weapons/nut_keys.lua:17]: And yet again here

File: plugins/doors/entities/weapons/nut_keys.lua [hunk 4]
```diff
@@ -1,164 +1,126 @@
...
 function SWEP:PreDrawViewModel(viewModel, weapon, client)
-	local hands = player_manager.TranslatePlayerHands(player_manager.TranslateToPlayerModelName(client:GetModel()))
-
-	if (hands and hands.model) then
-		--viewModel:SetModel(hands.model)
-		--viewModel:SetSkin(hands.skin)
-		--viewModel:SetBodyGroups(hands.body)
-	end
+    local hands = player_manager.TranslatePlayerHands(player_manager.TranslateToPlayerModelName(client:GetModel()))
+    if hands and hands.model then end --viewModel:SetModel(hands.model) --viewModel:SetSkin(hands.skin) --viewModel:SetBodyGroups(hands.body)
```
Comment (zoephix) [inline plugins/doors/entities/weapons/nut_keys.lua:40]: No need for a reformat.

File: plugins/doors/entities/weapons/nut_keys.lua [hunk 5]
```diff
@@ -1,164 +1,126 @@
...
+    local viewModel = self.Owner:GetViewModel()
+    if IsValid(viewModel) then end --viewModel:SetPlaybackRate(1) --viewModel:ResetSequence(ACT_VM_FISTS_DRAW)
 
-	if (IsValid(viewModel)) then
-		--viewModel:SetPlaybackRate(1)
-		--viewModel:ResetSequence(ACT_VM_FISTS_DRAW)
-	end
-
-	return true
+    return true
```
Comment (zoephix) [inline plugins/doors/entities/weapons/nut_keys.lua:51]: Also here, the format is fine as is.

File: plugins/doors/entities/weapons/nut_keys.lua [hunk 6]
```diff
@@ -1,164 +1,126 @@
...
-		viewModel:SetPlaybackRate(1)
-		viewModel:ResetSequence(ACT_VM_FISTS_HOLSTER)
-	end
+    if IsValid(viewModel) then
+        viewModel:SetPlaybackRate(1)
+        viewModel:ResetSequence(ACT_VM_FISTS_HOLSTER)
+    end
 
-	return true
+    return true
```
Comment (zoephix) [inline plugins/doors/entities/weapons/nut_keys.lua:63]: Again, no reformatting is needed.

File: plugins/doors/entities/weapons/nut_keys.lua [hunk 7]
```diff
@@ -1,164 +1,126 @@
...
+    self:SetNextSecondaryFire(CurTime() + time2)
+    if not IsFirstTimePredicted() then return end
+    if CLIENT then return end
+    local data = {}
+    data.start = self.Owner:GetShootPos()
+    data.endpos = data.start + self.Owner:GetAimVector() * 96
+    data.filter = self.Owner
+    local entity = util.TraceLine(data).Entity
+
+    --[[
```
Comment (zoephix) [inline plugins/doors/entities/weapons/nut_keys.lua:86]: Please don't, this formatting is terrible

File: plugins/doors/entities/weapons/nut_keys.lua [hunk 8]
```diff
@@ -1,164 +1,126 @@
...
-			(entity:IsVehicle() and entity.CPPIGetOwner and entity:CPPIGetOwner() == self.Owner)
-		)
-	) then
-		self.Owner:setAction("@locking", time, function()
-			self:toggleLock(entity, true)
-		end)			
+    if IsValid(entity) and ((entity:isDoor() and (entity:checkDoorAccess(self.Owner) or (self.Owner:IsAdmin() ))) or (entity:IsVehicle() and ((entity.CPPIGetOwner and entity:CPPIGetOwner() == self.Owner) or (entity.GetCreator and entity:GetCreator() == self.Owner)))) then
+        self.Owner:setAction("@locking", time, function()
+            self:toggleLock(entity, true)
+        end)
```
Comment (zoephix) [inline plugins/doors/entities/weapons/nut_keys.lua:94]: Remove redundant parentheses to simplify the conditional statements.
Indent the nested if statements for better readability.
Reformat the code to adhere to common Lua coding conventions.
Remove unnecessary spaces and align the code for consistency.

File: plugins/doors/entities/weapons/nut_keys.lua [hunk 9]
```diff
@@ -176,45 +138,28 @@ function SWEP:toggleLock(door, state)
...
+    local time2 = math.max(time, 1)
+    self:SetNextPrimaryFire(CurTime() + time2)
+    self:SetNextSecondaryFire(CurTime() + time2)
+    if not IsFirstTimePredicted() then return end
+    if CLIENT then return end
+    local data = {}
+    data.start = self.Owner:GetShootPos()
+    data.endpos = data.start + self.Owner:GetAimVector() * 96
+    data.filter = self.Owner
+    local entity = util.TraceLine(data).Entity
```
Comment (zoephix) [inline plugins/doors/entities/weapons/nut_keys.lua:151]: Again, please leave the formatting as is. You're not making it any better.

File: plugins/doors/entities/weapons/nut_keys.lua [hunk 10]
```diff
@@ -176,45 +138,28 @@ function SWEP:toggleLock(door, state)
...
-		return	
-	end
-end
+    if IsValid(entity) and ((entity:isDoor() and (entity:checkDoorAccess(self.Owner) or (self.Owner:IsAdmin() ))) or (entity:IsVehicle() and ((entity.CPPIGetOwner and entity:CPPIGetOwner() == self.Owner) or (entity.GetCreator and entity:GetCreator() == self.Owner)))) then
+        self.Owner:setAction("@unlocking", time, function()
+            self:toggleLock(entity, false)
+        end)
+
+        return
+    end
```
Comment (zoephix) [inline plugins/doors/entities/weapons/nut_keys.lua:164]: Remove redundant parentheses to simplify the conditional statements.
Indent the nested if statements for better readability.
Reformat the code to adhere to common Lua coding conventions.
Remove unnecessary spaces and align the code for consistency.

File: plugins/doors/sh_plugin.lua [hunk 11]
```diff
@@ -1,6 +1,6 @@
 PLUGIN.name = "Doors"
-PLUGIN.author = "Chessnut"
-PLUGIN.desc = "A simple door system."
+PLUGIN.author = "Chessnut & Leonheart#7476"
+PLUGIN.desc = "A simple door system // Now has multifaction support!"
```
Comment (zoephix) [inline plugins/doors/sh_plugin.lua:3]: No need to change the description.

File: plugins/doors/sh_plugin.lua [hunk 12]
```diff
@@ -11,41 +11,44 @@ nut.util.include("sv_plugin.lua")
...
+				
+				self.nutAccess = {}
+				self:SetDTEntity(0, nil)
 			end
-
-			self.nutAccess = {}
-			self:SetDTEntity(0, nil)
 		end
 	end
 end
```
Comment (zoephix) [inline plugins/doors/sh_plugin.lua:54]: I will ask why.

File: plugins/doors/sv_plugin.lua [hunk 13]
```diff
@@ -172,7 +157,13 @@ function PLUGIN:CanPlayerAccessDoor(client, door, access)
...
-	if (faction and client:Team() == faction) then
+	if(istable(faction)) then
+		for k, v in pairs(faction) do
+			if(client:Team() == v) then
+				return true
+			end
+		end
+	elseif (faction and client:Team() == faction) then
 		return true
 	end
```
Comment (zoephix) [inline plugins/doors/sv_plugin.lua:168]: You should consider accessing the factions by index to prevent having to iterate over them.

**Comment by bleonheart** [comment]
I will start prefacing this the same way I did with Tov. 

If you want to overcriticize, provide something that's useful towards your framework. All I hear about NutScript at this point is people coming to my account, regardless of the continuous allegations from some of your members, including you, which in the past harassed me on the content creation channel without any sort of provocation, asking me for help, since they get ridicularized on your discord by asking simple questions we all had. 

When I try to provide anything towards your community, which the Head of Staff made sure to enlighten that there's nothing specific against me, it's just a few guys saying I am a scammer and that I "groomed my wife" without any proof, I either get overcritized OR outright insulted.

You have had an massive exploit on your framework for years and when I tried to fix such FOR FREE, I got told to fuck off indirectly because "there's a better fix". That comes to show how much you care about your community. Ah, by the way, if you are a community member reading this, you can crash an server fairly easily by spam buying items with your inventory full. That will overload the server and make it implode.

Now, to finish this off, if you can do better, then do such, as you are an Developer for Nutscript (allegedly). I can't be arsed to read multiple critiques from a guy that provides nearly 0 to his own framework. In fact, I can go as far as saying that you spent more time doing this massive critique than providing for the framework in the past few months.

I've done my part and tried assisting the users of the framework, but if you want to be petty and act like a small child, be my guest. That shows why you had the need to fabricate so many false allegations about me and with this people will realise such.

**Comment by TovarischPootis** [comment]
Ad hominem insults and drama do not contribute to anything. The reviews were that of the code. Your outlash is childish at best

**Comment by zoephix** [comment]
Hi,

I understand you may be frustrated with the feedback I've provided on your pull request. However I feel that it's essential to reiterate that it's not my intention to overcriticize or insult you in any way, but rather to provide feedback.

As far as any claims made against you by other members of our community are concerned- It is crucially important to note here as an impartial moderator between reviewers and contributors not just harassment but any kind of false allegations should be looked into. I do want to emphasize that as developer, I don't condone any form of harrasment or allegations. I evaluate the technical basis and code quality, to ensure a secure codebase. I review any pull requests made based on that evaluation, not personal opinions or accusations.

I appreciate you providing assistance to others and offering a fix to what may be considered a major exploit. And although I value your contributions, it's important to keep in mind that feedback and suggestions from developers are intended to help maintain the integrity of the codebase.

Inadvertently if my feedback lacked clarity, then my apologies. It's not my intention to undermine your contributions in any way, I'm just ensuring the quality of the framework.

Thanks.

**Comment by bleonheart** [comment]
> Ad hominem insults and drama do not contribute to anything. The reviews were that of the code. Your outlash is childish at best

"Ad Hominem" my ass, I am just annoyed because Zoephnix has multiple instances of insulting me, as well as propagating "I only steal code".

I keep getting disrespected by you guys, which has been going for a while and, to be honest, I am getting a bit tired of that.  

In fact, I am yet to get ONE accusation with solid proof. 
I hear shit like I am a pedophile (grooming my 30yo wife surely must be nice), a scammer, a lua stealer and it most comes from NS.

So, if we want to get on talking about Ad Hominem, we can start on this. I made my peace of mind about this framework and I am currently working on my own, but I'd like to take a chance for people to see what type of behavior the NS staff had towards me. I got banned for "accusing the staff of pedophilia and scamming" which Falcon later on realized wasn't me doing. 
image
image
image
image
image
image
image
image

Adaster, which is widely known for his scams, and the perpetrator of the "staff are pedophiles" post (and was friends/staff on the server of a massive pedophile called BusterZlista which many of you will remember for posting CP on NS), was removed from the scammers list whilst I keep on being spread as a scamemr 
image

Adaster, which contributes NOTHING for the framework has scammed SEVERAL people on the past, only insults people and is generally hated, yet is on the server. Don't get me started on how I got unbanned in exchange for the ban of the real perpetrator of the (false) accusations and I got unbanned the day after, just for requesting a role.
image

Doing this rant right now, as I got several kids on my DMs getting scammed on NS and hearing all sorts of stuff about me. Hell I even got accused of making the rape swep lmfao. I am tired of having to deal with accusations myself whilst NS is full of scammers and the staff DO NOTHING about it.

The following screenshots were taken over an year and half apart. The guy is still in the server actively scamming kids.
image
image

So in conclusion, this is no "Ad Hominem". I am tired of having some individuals, which you both, especially Zoephnix, have their fair share, as seen above, spreading shit around me and then try to nitpick stuff I am providing FOR FREE.

All I am asking is that YOU STOP SPREADING MISINFORMATION like you've been for the several months. I can't be arsed about being unbanned but the things I hear that are originating from the NS discord are ridiculous. In fact, the community is dead for me for a while, but I want to get this out there, as I am sick and tired of hearing shit that comes from here.

And for those outside of the "Nutscript Lore" ask yourselves if you ever saw any actual proof regarding their accusations and how I'd be building an entire entire reputation with just "code stealing" (according to Zoephnix and many other). It's easier to ban the person you are unfairly abusing and mispreading info about than admitting errors of judgment. 

This convo ends my connection with this community, but I surely hope you do realize you've been attacking someone that's innocent with false accusations.

Either way, keep making NS great!

**Comment by TovarischPootis** [comment]
Duly noted. Thank you for your contributions

**Pull request closed without merging**